pkg/proxy/healthcheck/proxier_health.go

```diff
@@ -0,0 +1,161 @@
+	Updated()
+}
+
+// ProxierHealthServer returns 200 "OK" by default. It verifies that the delay between
+// QueuedUpdate() calls and Updated() calls never exceeds healthTimeout.
+type ProxierHealthServer struct {
+	listener    listener
+	httpFactory httpServerFactory
+	clock       clock.Clock
+
+	addr          string
+	port          int32
+	healthTimeout time.Duration
+	recorder      record.EventRecorder
+	nodeRef       *v1.ObjectReference
+
+	lastUpdated atomic.Value
+	lastQueued  atomic.Value
+}
+
+// NewProxierHealthServer returns a proxier health http server.
+func NewProxierHealthServer(addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) *ProxierHealthServer {
+	return newProxierHealthServer(stdNetListener{}, stdHTTPServerFactory{}, clock.RealClock{}, addr, healthTimeout, recorder, nodeRef)
+}
+
+func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration, recorder record.EventRecorder, nodeRef *v1.ObjectReference) *ProxierHealthServer {
+	return &ProxierHealthServer{
+		listener:      listener,
+		httpFactory:   httpServerFactory,
+		clock:         c,
+		addr:          addr,
+		healthTimeout: healthTimeout,
+		recorder:      recorder,
+		nodeRef:       nodeRef,
+	}
+}
+
+// Updated updates the lastUpdated timestamp.
+func (hs *ProxierHealthServer) Updated() {
+	hs.lastUpdated.Store(hs.clock.Now())
+}
+
+// QueuedUpdate updates the lastQueued timestamp.
+func (hs *ProxierHealthServer) QueuedUpdate() {
+	hs.lastQueued.Store(hs.clock.Now())
+}
+
+// Run starts the healthz http server and returns.
+func (hs *ProxierHealthServer) Run() {
+	serveMux := http.NewServeMux()
+	serveMux.Handle("/healthz", healthzHandler{hs: hs})
+	server := hs.httpFactory.New(hs.addr, serveMux)
+
+	go wait.Until(func() {
+		klog.V(3).Infof("Starting goroutine for proxier healthz on %s", hs.addr)
+
+		listener, err := hs.listener.Listen(hs.addr)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to start proxier healthz on %s: %v", hs.addr, err)
+			if hs.recorder != nil {
+				hs.recorder.Eventf(hs.nodeRef, api.EventTypeWarning, "FailedToStartProxierHealthcheck", msg)
+			}
+			klog.Error(msg)
+			return
+		}
+
+		if err := server.Serve(listener); err != nil {
+			klog.Errorf("Proxier healthz closed with error: %v", err)
+			return
+		}
+		klog.Error("Unexpected proxier healthz closed.")
+	}, proxierHealthzRetryInterval, wait.NeverStop)
+}
+
+type healthzHandler struct {
+	hs *ProxierHealthServer
+}
+
+func (h healthzHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	var lastQueued, lastUpdated time.Time
+	if val := h.hs.lastQueued.Load(); val != nil {
+		lastQueued = val.(time.Time)
+	}
+	if val := h.hs.lastUpdated.Load(); val != nil {
+		lastUpdated = val.(time.Time)
+	}
+	currentTime := h.hs.clock.Now()
+
+	healthy := false
+	switch {
+	case lastUpdated.IsZero():
+		// Consider the proxy healthy while it's starting up
```

Comment:
I think we want to do the exact opposite - until the first Updated, we're not healthy.

This is a behavior change, though.

Reply:
Yeah, I was trying to change the behavior as little as possible for now

Reply:
TOOD ?

Reply:
added a comment